File: windows/useragent.test.js
// #134 tests:  node windows/useragent.test.js
const assert = require('assert');
const { cleanUserAgent } = require('./useragent');

let run = 0;
const test = (name, fn) => {
  fn();
  run++;
  console.log(`  ok  ${name}`);
};

// The real thing, copied from Electron 34.5.8 / Chromium 132 on Windows.
const REAL =
  'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) ' +
  'WebForge/0.1.142 Chrome/132.0.0.0 Electron/34.5.8 Safari/537.36';

console.log('we stop advertising that we are Electron');

test('both giveaway tokens are gone', () => {
  const ua = cleanUserAgent(REAL);
  assert.ok(!/Electron/i.test(ua), `Electron token survived: ${ua}`);
  assert.ok(!/WebForge/i.test(ua), `WebForge token survived: ${ua}`);
});

test('the result is exactly what Chrome would send', () => {
  assert.strictEqual(
    cleanUserAgent(REAL),
    'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) ' +
      'Chrome/132.0.0.0 Safari/537.36'
  );
});

console.log('the parts that must survive, survive');

test('the Chrome version token is untouched — sniffers gate on it', () => {
  assert.ok(cleanUserAgent(REAL).includes('Chrome/132.0.0.0'));
});

test('platform, AppleWebKit and Safari tokens are untouched', () => {
  const ua = cleanUserAgent(REAL);
  for (const part of [
    '(Windows NT 10.0; Win64; x64)',
    'AppleWebKit/537.36',
    '(KHTML, like Gecko)',
    'Safari/537.36',
  ]) {
    assert.ok(ua.includes(part), `lost ${part}`);
  }
});

console.log('the string never comes out malformed');

test('no double spaces where the tokens were removed', () => {
  assert.ok(!/ {2,}/.test(cleanUserAgent(REAL)), 'double space betrays the edit');
});

test('no leading or trailing whitespace', () => {
  assert.strictEqual(cleanUserAgent(`  ${REAL}  `), cleanUserAgent(REAL));
});

console.log('it holds up on inputs that are not the happy path');

test('a UA with neither token is returned unchanged', () => {
  const plain =
    'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) ' +
    'Chrome/132.0.0.0 Safari/537.36';
  assert.strictEqual(cleanUserAgent(plain), plain);
});

test('running it twice changes nothing', () => {
  const once = cleanUserAgent(REAL);
  assert.strictEqual(cleanUserAgent(once), once);
});

test('a token at the very end is still removed', () => {
  // Chromium has moved token order between versions; do not assume Safari is last.
  assert.strictEqual(
    cleanUserAgent('Mozilla/5.0 Chrome/132.0.0.0 Safari/537.36 Electron/34.5.8'),
    'Mozilla/5.0 Chrome/132.0.0.0 Safari/537.36'
  );
});

test('a differently-cased app token is still removed', () => {
  assert.ok(!/webforge/i.test(cleanUserAgent('Mozilla/5.0 webforge/1.2.3 Chrome/132.0.0.0')));
});

test('junk in does not throw', () => {
  for (const junk of [undefined, null, '', 0, {}, []]) {
    assert.strictEqual(cleanUserAgent(junk), '', `threw or leaked on ${JSON.stringify(junk)}`);
  }
});

console.log(`\n${run} tests passed`);

File: android/app/src/test/java/com/webforge/browser/UserAgentTest.kt
package com.webforge.browser

import org.junit.Assert.assertEquals
import org.junit.Assert.assertFalse
import org.junit.Assert.assertTrue
import org.junit.Test

/**
 * #134: the phone must stop identifying as an embedded WebView, for the same
 * reason the desktop stops identifying as Electron. Runs on the JVM (no device)
 * because [UserAgent] is free of Android APIs — same pattern as [VaultCryptoTest].
 *
 * The Windows half of this lives in `windows/useragent.test.js`; the two assert
 * the same properties against their own engine's string.
 */
class UserAgentTest {

    // A real Android System WebView UA (Chrome 132 era).
    private val real =
        "Mozilla/5.0 (Linux; Android 14; Pixel 7 Build/AP1A.240505.004; wv) " +
            "AppleWebKit/537.36 (KHTML, like Gecko) Version/4.0 Chrome/132.0.0.0 " +
            "Mobile Safari/537.36"

    @Test
    fun `both WebView giveaway tokens are gone`() {
        val ua = UserAgent.clean(real)
        assertFalse("wv token survived: $ua", Regex("""\bwv\b""").containsMatchIn(ua))
        assertFalse("Version token survived: $ua", ua.contains("Version/"))
    }

    @Test
    fun `the result is exactly what mobile Chrome would send`() {
        assertEquals(
            "Mozilla/5.0 (Linux; Android 14; Pixel 7 Build/AP1A.240505.004) " +
                "AppleWebKit/537.36 (KHTML, like Gecko) Chrome/132.0.0.0 " +
                "Mobile Safari/537.36",
            UserAgent.clean(real)
        )
    }

    @Test
    fun `the Chrome version token is untouched - sniffers gate on it`() {
        assertTrue(UserAgent.clean(real).contains("Chrome/132.0.0.0"))
    }

    @Test
    fun `device and platform details survive`() {
        val ua = UserAgent.clean(real)
        for (part in listOf("Linux; Android 14", "Pixel 7", "AppleWebKit/537.36", "Mobile Safari/537.36")) {
            assertTrue("lost $part", ua.contains(part))
        }
    }

    @Test
    fun `no ragged spacing is left behind`() {
        val ua = UserAgent.clean(real)
        assertFalse("double space betrays the edit", ua.contains("  "))
        assertFalse("dangling separator", ua.contains("; )"))
        assertFalse("dangling separator", ua.contains(" )"))
    }

    @Test
    fun `a UA with neither token is returned unchanged`() {
        val plain =
            "Mozilla/5.0 (Linux; Android 14) AppleWebKit/537.36 (KHTML, like Gecko) " +
                "Chrome/132.0.0.0 Mobile Safari/537.36"
        assertEquals(plain, UserAgent.clean(plain))
    }

    @Test
    fun `running it twice changes nothing`() {
        val once = UserAgent.clean(real)
        assertEquals(once, UserAgent.clean(once))
    }

    @Test
    fun `a UA that only says wv is still handled`() {
        assertEquals(
            "Mozilla/5.0 (Linux; Android 14) Chrome/132.0.0.0",
            UserAgent.clean("Mozilla/5.0 (Linux; Android 14; wv) Chrome/132.0.0.0")
        )
    }

    @Test
    fun `junk in does not throw`() {
        assertEquals("", UserAgent.clean(null))
        assertEquals("", UserAgent.clean(""))
    }
}
